File: src/lib/agent/apply.ts
import { escapeHtml } from "../resume-render";
import { sendMail, type SendResult } from "../mail";
import type { Application, ApplicationChannel, Job, Profile } from "../types";

export function channelFor(job: Job): ApplicationChannel {
  return job.applyEmail ? "email" : "external_form";
}

function subjectFor(job: Job, profile: Profile): string {
  return `Application: ${job.title} — ${profile.fullName}`;
}

function textBody(job: Job, profile: Profile, application: Pick<Application, "coverLetter" | "resumeMarkdown">): string {
  return [
    application.coverLetter,
    "",
    "—".repeat(20),
    "",
    `Resume — ${profile.fullName}`,
    "",
    application.resumeMarkdown,
    "",
    `Posting reference: ${job.url}`,
  ].join("\n");
}

function htmlBody(job: Job, profile: Profile, application: Pick<Application, "coverLetter" | "resumeMarkdown">): string {
  const letter = application.coverLetter
    .split(/\n{2,}/)
    .map((paragraph) => `<p style="margin:0 0 14px">${escapeHtml(paragraph).replace(/\n/g, "<br />")}</p>`)
    .join("\n");

  return `<div style="font-family:-apple-system,Segoe UI,Helvetica,Arial,sans-serif;font-size:15px;line-height:1.6;color:#111827;max-width:640px">
${letter}
<hr style="border:none;border-top:1px solid #e5e7eb;margin:24px 0" />
<p style="margin:0 0 6px;font-size:13px;color:#6b7280">Resume attached as Markdown. Posting reference: <a href="${escapeHtml(job.url)}">${escapeHtml(job.url)}</a></p>
<p style="margin:0;font-size:13px;color:#6b7280">${escapeHtml(profile.fullName)} · ${escapeHtml(profile.email)}${profile.phone ? ` · ${escapeHtml(profile.phone)}` : ""}</p>
</div>`;
}

export interface SubmitOutcome {
  submitted: boolean;
  result: SendResult | null;
  message: string;
}

/**
 * Sends an email application for postings that advertise an address. Postings
 * behind an ATS form are never auto-filled: scripted form submission violates
 * most job boards' terms of service, so those are handed back for review.
 */
export async function submitByEmail(
  job: Job,
  profile: Profile,
  application: Pick<Application, "coverLetter" | "resumeMarkdown">,
): Promise<SubmitOutcome> {
  if (!job.applyEmail) {
    return {
      submitted: false,
      result: null,
      message: "This posting has no application address, so it has to be submitted through its own form.",
    };
  }

  const slug = `${profile.fullName.replace(/\s+/g, "-")}-${job.company.replace(/\s+/g, "-")}`.toLowerCase();
  const result = await sendMail({
    to: job.applyEmail,
    subject: subjectFor(job, profile),
    text: textBody(job, profile, application),
    html: htmlBody(job, profile, application),
    attachments: [
      {
        filename: `${slug}-resume.md`,
        content: application.resumeMarkdown,
        contentType: "text/markdown",
      },
    ],
  });

  if (result.transport === "smtp" && result.ok) {
    return { submitted: true, result, message: `Emailed to ${job.applyEmail}. ${result.detail}` };
  }

  return {
    submitted: false,
    result,
    message: result.ok
      ? `Application drafted but not delivered: ${result.detail}`
      : `Application could not be delivered: ${result.error ?? result.detail}`,
  };
}

File: app/apply.py
"""Decide how an application leaves the machine.

Email postings can be sent (or written to the outbox). Greenhouse / Lever /
Workday / Easy Apply forms are never filled in by this agent — scripted
submission violates most boards' terms of service. Those get a prepared pack
and a URL.
"""

from __future__ import annotations

from app.mail import MailAttachment, MailMessage, SendResult, send_mail
from app.models import Job, Profile, TailoredApplication
from app.resume_render import escape_html


def channel_for(job: Job) -> str:
    return "email" if job.apply_email else "external_form"


def _subject(job: Job, profile: Profile) -> str:
    return f"Application: {job.title} — {profile.full_name}"


def _text(job: Job, profile: Profile, tailored: TailoredApplication) -> str:
    return "\n".join(
        [
            tailored.cover_letter,
            "",
            "—" * 20,
            "",
            f"Resume — {profile.full_name}",
            "",
            tailored.resume_markdown,
            "",
            f"Posting reference: {job.url}",
        ]
    )


def _html(job: Job, profile: Profile, tailored: TailoredApplication) -> str:
    letter = "".join(
        f'<p style="margin:0 0 14px">{escape_html(paragraph).replace(chr(10), "<br />")}</p>'
        for paragraph in tailored.cover_letter.split("\n\n")
    )
    return f"""<div style="font-family:-apple-system,Segoe UI,Helvetica,Arial,sans-serif;font-size:15px;line-height:1.6;max-width:640px">
{letter}
<p style="font-size:13px;color:#6b7280">Posting: <a href="{escape_html(job.url)}">{escape_html(job.url)}</a></p>
</div>"""


def submit_by_email(job: Job, profile: Profile, tailored: TailoredApplication) -> tuple[bool, SendResult | None, str]:
    if not job.apply_email:
        return (
            False,
            None,
            "This posting has no application address; submit through the company's own form.",
        )

    slug = f"{profile.full_name}-{job.company}".replace(" ", "-").lower()
    result = send_mail(
        MailMessage(
            to=job.apply_email,
            subject=_subject(job, profile),
            text=_text(job, profile, tailored),
            html=_html(job, profile, tailored),
            attachments=[
                MailAttachment(
                    filename=f"{slug}-resume.md",
                    content=tailored.resume_markdown,
                    content_type="text/markdown",
                )
            ],
        )
    )
    if result.transport == "smtp" and result.ok:
        return True, result, f"Emailed to {job.apply_email}. {result.detail}"
    return (
        False,
        result,
        (
            f"Application drafted but not delivered: {result.detail}"
            if result.ok
            else f"Application could not be delivered: {result.error or result.detail}"
        ),
    )
